File: iosApp/iosApp/feature/consumer_cart/ConsumerCartViewState.swift
//
//  ConsumerCartViewState.swift
//  iosApp
//
//  Created by Марк Шавловский on 17.05.2024.
//  Copyright © 2024 orgName. All rights reserved.
//

import Foundation

struct ConsumerCartViewState {
    var state: ConsumerCartState
}

enum ConsumerCartState{
    case loading
    case error
    case success([CartProductItemUi], [MenuProductItem], BottomPanelInfoUi?)
}

struct CartProductItemUi: Identifiable {
    var id: String
    var name: String
    var newCost: String
    var oldCost: String?
    var photoLink: String
    var count: Int
    var additions: String?
    var isLast: Bool
}

struct BottomPanelInfoUi {
    var motivation: MotivationUi?
    var discount: String?
    var oldTotalCost: String?
    var newTotalCost: String
}

enum MotivationUi {
    case MinOrderCost(String)
    case ForLowerDelivery(String, Float, Bool)
    case LowerDeliveryAchieved(Bool)
}

File: feature/consumercart/src/commonMain/kotlin/com/bunbeauty/order/ui/state/ConsumerCartViewState.kt
package com.bunbeauty.order.ui.state

import androidx.compose.runtime.Immutable
import com.bunbeauty.core.base.BaseViewState
import com.bunbeauty.core.motivation.MotivationUi
import kotlinx.collections.immutable.ImmutableList

@Immutable
sealed class ConsumerCartViewState(
    val state: State,
) : BaseViewState {
    @Immutable
    data object Loading : ConsumerCartViewState(state = State.LOADING)

    @Immutable
    data class Success(
        val cartProductList: ImmutableList<CartProductItemUi>,
        val recommendationList: ImmutableList<ProductUi>,
        val bottomPanelInfo: BottomPanelInfoUi?,
    ) : ConsumerCartViewState(state = State.SUCCESS) {
        data class ProductUi(
            val key: String,
            val uuid: String,
            val photoLink: String,
            val name: String,
            val oldPrice: String?,
            val newPrice: String,
        )
    }

    @Immutable
    data object Error : ConsumerCartViewState(state = State.ERROR)

    @Immutable
    data class BottomPanelInfoUi(
        val motivation: MotivationUi?,
        val discount: String?,
        val oldTotalCost: String?,
        val newTotalCost: String,
        val orderAvailable: Boolean,
    )

    @Immutable
    data class CartProductItemUi(
        val key: String,
        val uuid: String,
        val name: String,
        val newCost: String,
        val oldCost: String?,
        val photoLink: String,
        val count: Int,
        val additions: String?,
        val isLast: Boolean,
    )

    enum class State {
        LOADING,
        ERROR,
        SUCCESS,
    }
}
